### PR TITLE
G-API: Refine Semantic Segmentation Demo

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Overview
* Supported demo working with camera id (e.g `--input=0`)
* Supported 3d output segmentation models (e.g `deeplabv3`)
* Supported `desync` execution
* Supported higher camera resolution
* Changed the color map to pascal voc (https://cloud.githubusercontent.com/assets/4503207/17803328/1006ca80-65f6-11e6-9ff6-36b7ef5b9ac6.png)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/streaming/cap.hpp [hunk 1]
```diff
@@ -47,8 +47,17 @@ namespace wip {
 class GCaptureSource: public IStreamSource
 {
 public:
-    explicit GCaptureSource(int id) : cap(id) { prep(); }
-    explicit GCaptureSource(const std::string &path) : cap(path) { prep(); }
+    using Properties = std::unordered_map<int, double>;
+    explicit GCaptureSource(int id, const Properties &properties = {})
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/cap.hpp:51]: Well, I wish it was separate method `set` but properties must be set before the first frame is read (`prep()`)

Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/streaming/cap.hpp:51]: Not exposed to python so far...

File: modules/gapi/samples/semantic_segmentation.cpp [hunk 2]
```diff
@@ -5,34 +5,23 @@
...
-    { 100, 60,  0 },
-    { 90,  0,   0 },
-    { 230, 0,   0 },
-    { 32,  11,  119 },
-    { 0,   74,  111 },
+static std::vector<cv::Vec3b> colors = {
+    {0, 0, 0},       {0, 0, 128},     {0, 128, 0}, {0, 128, 128}, {128, 0, 0},   {128, 0, 128}, {128, 128, 0},
+    {128, 128, 128}, {0, 0, 64},      {0, 0, 192}, {0, 128, 64},  {0, 128, 192}, {128, 0, 64},  {128, 0, 192},
+    {128, 128, 64},  {128, 128, 192}, {0, 64, 0},  {0, 64, 128},  {0, 192, 0},   {0, 192, 128}, {128, 64, 0}
 };
```
Comment (dmatveev) [inline modules/gapi/samples/semantic_segmentation.cpp:25]: What's the reason for this change? Previously it was just one line per class, easy to edit.

File: modules/gapi/samples/semantic_segmentation.cpp [hunk 3]
```diff
@@ -97,6 +97,25 @@ void probsToClasses(const cv::Mat& probs, cv::Mat& classes) {
...
+    cv::Scalar color = {200, 10, 10};
+    double fontScale = 0.65;
+
+    cv::putText(mat, message, position, fontFace,
+                fontScale, cv::Scalar(255, 255, 255), thickness + 1);
+    cv::putText(mat, message, position, fontFace, fontScale, color, thickness);
+}
+
+static void drawResults(cv::Mat &img, const cv::Mat &color_mask) {
+    img = img / 2 + color_mask / 2;
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/samples/semantic_segmentation.cpp:114]: Probably we should thing how we can make stuff like this in graph -- where a desynchronized path meets the synchronized one, and results are accumulated inside as we do now manually..

Good idea for future versions

**Review by dmatveev — APPROVED** [review]

File: modules/gapi/samples/semantic_segmentation.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/samples/semantic_segmentation.cpp:25]: No problem, I'll keep old alignment

**Comment by TolyaTalamanov** [comment]
@dmatveev @asmorkalov Hi folks, can we merge this?

File: modules/gapi/samples/semantic_segmentation.cpp [hunk 4]
```diff
@@ -145,40 +198,88 @@ int main(int argc, char *argv[]) {
...
+    std::shared_ptr<cv::gapi::wip::GCaptureSource> source;
+    if (isNumber(input)) {
+        source = std::make_shared<cv::gapi::wip::GCaptureSource>(
+            std::stoi(input),
+            std::map<int, double> {
+              {cv::CAP_PROP_FRAME_WIDTH, 1280},
+              {cv::CAP_PROP_FRAME_HEIGHT, 720},
+              {cv::CAP_PROP_BUFFERSIZE, 1},
+              {cv::CAP_PROP_AUTOFOCUS, true},
+              {cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G')}
```
Comment (asmorkalov) [inline modules/gapi/samples/semantic_segmentation.cpp:219]: It works not with all cameras. I propose to add note on required camera parameters to help message and the source file header.

File: modules/python/src2/cv2_convert.hpp [hunk 5]
```diff
@@ -268,6 +268,11 @@ PyObject* pyopencv_from(const std::vector<Tp>& value)
 template<typename K, typename V>
 bool pyopencv_to(PyObject *obj, std::map<K,V> &map, const ArgInfo& info)
 {
+    if (!obj || obj == Py_None)
+    {
+        return true;
+    }
```
Comment (asmorkalov) [inline modules/python/src2/cv2_convert.hpp:274]: @VadimLevin Is is relevant patch?

File: modules/gapi/samples/semantic_segmentation.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/samples/semantic_segmentation.cpp:219]: Good point, I'd rather remove it then

File: modules/python/src2/cv2_convert.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/python/src2/cv2_convert.hpp:274]: I guess it's the common check that I missed. It's used for all other wrappers.

The code below won't work without this check.
```
GAPI_EXPORTS_W void foo(const map<int, int> &map = {});
```

File: modules/gapi/samples/semantic_segmentation.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/samples/semantic_segmentation.cpp]: Done
